include/flucoma/clients/nrt/KDTreeClient.hpp

```diff
@@ -188,74 +194,74 @@ class KDTreeQuery : public FluidBaseClient, ControlIn, ControlOut
 
       index k = get<kNumNeighbors>();
       if (k > kdtreeptr->size() || k < 0)
         return; // c.reportError("FluidKDTree RT Query has wrong k size");
-      index             dims = kdtreeptr->dims();
+
+      index dims = kdtreeptr->dims();
+
       InOutBuffersCheck bufCheck(dims);
       if (!bufCheck.checkInputs(get<kInputBuffer>().get(),
                                 get<kOutputBuffer>().get()))
         return; // c.reportError("FluidKDTree RT Query i/o buffers are
                 // unavailable");
-      auto datasetClientPtr = get<kDataSet>().get().lock();
-      if (!datasetClientPtr)
-        datasetClientPtr = kdtreeptr->getDataSet().get().lock();
 
-      if (!datasetClientPtr)
-      {
-        // c.reportError("Could not obtain reference FluidDataSet");
-        return;
-      }
+      auto lookupDSpointer = get<kLookupDataSet>().get().lock();
 
-      auto  dataset = datasetClientPtr->getDataSet();
-      index pointSize = dataset.pointSize();
-      auto  outBuf = BufferAdaptor::Access(get<kOutputBuffer>().get());
-      index maxK = outBuf.samps(0).size() / pointSize;
-      if (maxK <= 0) return;
-      index outputSize = maxK * pointSize;
+      index pointSize = lookupDSpointer ? lookupDSpointer->dims().value() : 1;
+
+      auto outBuf = BufferAdaptor::Access(get<kOutputBuffer>().get());
+      auto outSamps = outBuf.samps(0);
+
+      index numPoints = outSamps.size() / pointSize;
+      if (numPoints <= 0)
+        return; // c.reportError("FluidKDTree RT Query output buffer is too
+                // small for one point")
 
       RealVector point(dims, c.allocator());
       point <<= BufferAdaptor::ReadAccess(get<kInputBuffer>().get())
                     .samps(0, dims, 0);
-      if (mRTBuffer.size() != outputSize)
-      {
-        mRTBuffer = RealVector(outputSize, c.allocator());
-        mRTBuffer.fill(0);
-      }
 
       auto [dists, ids] = kdtreeptr->algorithm().kNearest(
           point, k, get<kRadius>(), c.allocator());
 
-      mNumValidKs = std::min(asSigned(ids.size()), maxK);
+      if (lookupDSpointer)
+      {
+        auto lookupDS = lookupDSpointer->getDataSet();
+
+        auto lookupFn = [&lookupDS, outSamps, pointSize,
+                         n = 0](auto id) mutable {
+          if (auto point = lookupDS.get(*id); point.data() != nullptr)
+            outSamps(Slice(n, pointSize)) <<= point;
+          n += pointSize;
+        };
 
-      for (index i = 0; i < mNumValidKs; i++)
+        std::for_each_n(ids.begin(), numPoints, lookupFn);
+      }
+      else
       {
-        dataset.get(*ids[asUnsigned(i)],
-                    mRTBuffer(Slice(i * pointSize, pointSize)));
+        std::transform(dists.begin(), dists.begin() + numPoints,
+                       outSamps.begin(), [](auto p) { return std::sqrt(p); });
       }
-      outBuf.samps(0, outputSize, 0) <<= mRTBuffer;
+
+      mLastNumPoints = ids.size();
```

Comment:
To be consistent with the old code, shouldn't this be `min(ids.size(), numPoints)`?

Reply:
indeed, sorted